### ISSUE TITLE
Breadcrumb title does not work display news title

### ISSUE THREAD

**Issue by bennyborn** [contao/contao#6050]
### Affected version(s)

^4.13

### Description

We like using `RootPageDependentModulesController` to setup e.g. `NewsReader` modules for multi-language installations since it save us a bit of time and keeps the backend a bit cleaner.

Recently we noticed a strange behaviour using this module in combination with `NewsReader` modules (maybe also others).

Normally a `NewsReader` overwrites the meta title of the current page with the data from the news entry itself. This is also shown in the breadcrumb navigation, so the active page isn't called e.g. `Details` anymore rather than `My awesome news article`. This works absolutely fine if you just add a reader module into your article on your detail page.

But if you decide to use the `RootPageDependentModulesController` instead, the active page in the breadcrumb will still have its original title instead the meta title of the news entry.

**Comment by aschempp** [comment]
Very interesting, that does not make sense to me (without inspecting the code). Have you tried adding the breadcrumb module through an insert tag (e.g. using `{{insert_module::...}}` in an HTML module? I usually have the issue that the breadcrumb is generated before the reader module is executed, because the breadcrumb is in the header. But that should not be related to the `RootPageDependentModulesController`.

**Comment by bennyborn** [comment]
> Have you tried adding the breadcrumb module through an insert tag (e.g. using `{{insert_module::...}}` in an HTML module?

Nope, but it seems to me that the `RootPageDependentModulesController` pretty much behaves the same as an insert tag.

> I usually have the issue that the breadcrumb is generated before the reader module is executed, because the breadcrumb is in the header.

That, on the other side, has never posed a problem for me. Of course the breadcrumb is mostly located before the reader but so far it has always worked without a problem for me 🤷‍♂️

**Comment by fritzmg** [comment]
> But if you decide to use the `RootPageDependentModulesController` instead, the active page in the breadcrumb will still have its original title instead the meta title of the news entry.

So when you insert the breadcrumb module _directly_ (instead via the root page dependent modules module) this issue does not occur?

> Nope, but it seems to me that the `RootPageDependentModulesController` pretty much behaves the same as an insert tag.

No, an insert tag is replaced later in the pipeline, especially one added in the layout. (For Contao 5 it is slightly different, but should still be the same for insert tags in the page layout.)

> That, on the other side, has never posed a problem for me. Of course the breadcrumb is mostly located before the reader but so far it has always worked without a problem for me 🤷‍♂️

That cannot be the case as at that point the breadcrumb module does not have the updated meta data information yet.


In any case - yes, this has always been a problem and is not related to the `RootPageDependentModulesController`. We need a more general solution so that modules like `breadcrumb` accesses the `HtmlHeadBag` once all relevant modules have updated the information in the `HtmlHeadBag`. i.e. its rendering always needs to be deferred (rendering as `esi` is likely also not a solution since the fragment will not have access to the `HtmlHeadBag` of the originating master request - presumably).

**Comment by bennyborn** [comment]
> So when you insert the breadcrumb module _directly_ (instead via the root page dependent modules module) this issue does not occur?

You misunderstood me there.
The breadcrumb module is always "bare". I'm using the `RootPageDependentModulesController` for the `NewsReaderModule`.

> That cannot be the case as at that point the breadcrumb module does not have the updated meta data information yet.

Then I just got lucky or a bug all these years? 
This "article layout" always worked fine so far (tbh it seems broken in the 5.1 Online Demo too).

```
--------------------------------
|          BREADCRUMB          |
--------------------------------
|                              |
|          NEWS READER         |
|                              |
--------------------------------
```

**Comment by fritzmg** [comment]
It simply depends on the order of modules being processed. If the breadcrumb module is executed before the newsreader module, then the breadcrumb module will naturally not have the updated information in the `HtmlHeadBag`. You can debug in your setups which modules are processed first and why.

**Comment by aschempp** [comment]
didn't we add the current page title as an insert tag a few years ago to work around this issue?

**Comment by fritzmg** [comment]
If we did, I can't find any evidence of that. This wouldn't work for Contao 5 anyway, as the insert tag would be replaced to soon.

I think the correct fix would be to implement a `NewsDetailPageController` which will fill the `HtmlHeadBag` with the correct information before any content element or module is rendered.

**Comment by aschempp** [comment]
yeah, that's exactly what custom page controllers are design for 😆 but we don't have that yet – and not sure we want to add that for legacy extensions? If we do, I'm happy to start a draft PR 😊

**Comment by fritzmg** [comment]
See https://github.com/contao/contao/issues/7105#issuecomment-2704189223 for a solution until #8085 is implemented.

**Comment by fritzmg** [comment]
As discussed in the Contao call we want to introduce a `getName()`/`setName()` on the `HtmlHeadBag`, which can be set by a reader page (or a reader module) then be read by `ModuleBreadcrumb` for the active item.
